### PR TITLE
Move random into Suite

### LINKED ISSUE #257: remove random.Stream
...and make the heavy API changes necessary so that all places it was used can specify a seeded RNG if they need stability during testing.

See #253 for the hacky way we make this possible today, that we want to remove for v1, hopefully.

### PR DESCRIPTION
**Pull request by jeffallen**
Before randomness was passed in via arguments, or sometimes
taken from the global random.Stream. With this change, randomness
comes from the Suite, simplifying the APIs and also making it
possible for tests to elect to have predictable random numbers.

Closes #257.

### CODE REVIEW

File: group/curve25519/curve.go [hunk 1]
```diff
@@ -391,3 +388,16 @@ func (c *curve) data(x, y *mod.Int) ([]byte, error) {
 	}
 	return b[1 : 1+dl], nil
 }
+
+// reverse copies src into dst in byte-reversed order and returns dst,
+// such that src[0] goes into dst[len-1] and vice versa.
+// dst and src may be the same slice but otherwise must not overlap.
+func reverse(dst, src []byte) []byte {
+	l := len(dst)
+	for i, j := 0, l-1; i < (l+1)/2; {
```
Comment (ineiti) [inline group/curve25519/curve.go]: code:
```
for i, j := 0, l-1; i < (l+1)/2; i++, j--
```

File: group/edwards25519/scalar.go [hunk 2]
```diff
@@ -2231,3 +2230,16 @@ func scReduce(out *[32]byte, s *[64]byte) {
 	out[30] = byte(s11 >> 9)
 	out[31] = byte(s11 >> 17)
 }
+
+// reverse copies src into dst in byte-reversed order and returns dst,
+// such that src[0] goes into dst[len-1] and vice versa.
+// dst and src may be the same slice but otherwise must not overlap.
+func reverse(dst, src []byte) []byte {
+	l := len(dst)
+	for i, j := 0, l-1; i < (l+1)/2; {
```
Comment (ineiti) [inline group/edwards25519/scalar.go]: code:
```
for i, j := 0, l-1; i < (l+1)/2; i++, j--
```

File: group/mod/int.go [hunk 3]
```diff
@@ -469,3 +468,19 @@ func (i *Int) HideDecode(buf []byte) {
 	i.V.SetBytes(buf)
 	i.V.Mod(&i.V, i.M)
 }
+
+// reverse copies src into dst in byte-reversed order and returns dst,
+// such that src[0] goes into dst[len-1] and vice versa.
+// dst and src may be the same slice but otherwise must not overlap.
+func reverse(dst, src []byte) []byte {
```
Comment (ineiti) [inline group/mod/int.go]: Now we use it three times - might be worth exporting it here, no?

File: group/curve25519/curve.go (hunk 1 above)
Comment (jeffallen) [reply group/curve25519/curve.go]: That's not vaild Go. Only one expression is allowed after the second semi-colon. (surprised me too!)

File: group/mod/int.go (hunk 3 above)
Comment (jeffallen) [reply group/mod/int.go]: exporting into which package? into util/bytes? The point of kyber is to provide crypto primitives, not byte reversing. A new reader exploring the API should find things in util/ that are related to crypto, like encoding scalars and points, making keys. Not reversing bytes.

**Comment by jeffallen** [comment]
The way to override is implemented only in Edwards25519 curves for the moment: edwards25519.NewBlakeSHA256Ed25519WithRand(blake.New(nil)).

It is used in sign/anon/enc_test.go.

I will call more attention to it when I go through the docs this week.

File: group/curve25519/curve.go (hunk 1 above)
Comment (ineiti) [reply group/curve25519/curve.go]: Now I see how they did it at https://github.com/golang/go/wiki/SliceTricks :
```
for i, j := 0, l-1; i < (l+1)/2; i, j = i+1, j-1
```

**Pull request merged**